File: backend/app/services/contract.py
import os
import docx

TEMPLATE_PATH = r"c:\Projetos\Cosmo Alma TV\portal-egregora\Contrato\CONTRATO_COSMO_ALMA_TV_V3.docx"
OUTPUT_DIR = r"c:\Projetos\Cosmo Alma TV\portal-egregora\backend\generated_contracts"

def generate_contract_docx(
    nome_completo: str, 
    nome_comercial: str, 
    razao_social: str, 
    cnpj_cpf: str, 
    email: str, 
    chave_pix: str, 
    youtube_id: str,
    genero: str = None,
    estado_civil: str = None,
    cep: str = None,
    endereco: str = None,
    cidade: str = None,
    uf: str = None,
    pais: str = None
) -> str:
    """
    Generates a personalized contract based on the template by replacing the default values.
    Returns the path to the generated contract file.
    """
    if not os.path.exists(OUTPUT_DIR):
        os.makedirs(OUTPUT_DIR)

    doc = docx.Document(TEMPLATE_PATH)

    # Gender adjustments
    genero_val = (genero or "").upper()
    nacionalidade = "brasileiro(a)"
    if genero_val == "M":
        nacionalidade = "brasileiro"
    elif genero_val == "F":
        nacionalidade = "brasileira"

    # Address block
    address_str = f"{endereco or ''}, {cidade or ''} - {uf or ''}, CEP {cep or ''}, {pais or 'Brasil'}"

    # Determine tags
    clean_doc = "".join([c for c in cnpj_cpf if c.isdigit()])
    is_cnpj = len(clean_doc) > 11
    tipo_pessoa = "jurídica" if is_cnpj else "física"
    cpf_ou_cnpj = "CNPJ" if is_cnpj else "CPF"
    nome_parte = razao_social or nome_completo

    # Replacements dictionary based on V3 bracketed placeholders
    replacements = {
        "[nome PESSOA FISICA OU JURÍDICA]": nome_parte.upper(),
        "[física/jurídica]": tipo_pessoa,
        "[CPF/CNPJ]": cpf_ou_cnpj,
        "[número cpf OU cnpj]": cnpj_cpf,
        "[NOME]": nome_completo.upper(),
        "[ESTADO civil]": (estado_civil or "solteiro(a)").lower(),
        "[CPF]": cnpj_cpf,
        "[ENDEREÇO]": address_str,
    }

    # Helper function to replace text in runs while preserving style
    def replace_text_in_paragraphs(paragraphs):
        for paragraph in paragraphs:
            for key, value in replacements.items():
                if key in paragraph.text:
                    # Replace in the whole paragraph text and reset runs to avoid breaking split text
                    text = paragraph.text.replace(key, value)
                    paragraph.text = text

    # Replace in main body paragraphs
    replace_text_in_paragraphs(doc.paragraphs)

    # Replace in tables
    for table in doc.tables:
        for row in table.rows:
            for cell in row.cells:
                replace_text_in_paragraphs(cell.paragraphs)

    # Save to generated contracts
    clean_cnpj_cpf = cnpj_cpf.replace('/', '').replace('.', '').replace('-', '')
    filename = f"contrato_{nome_comercial.lower().replace(' ', '_')}_{clean_cnpj_cpf}.docx"
    output_path = os.path.join(OUTPUT_DIR, filename)
    doc.save(output_path)
    return output_path

File: src/lib/services/contract.ts
import fs from "fs";
import path from "path";
import PizZip from "pizzip";

// We will look for the template in the root of the project
const TEMPLATE_RELATIVE_PATH = "Contrato/CONTRATO_COSMO_ALMA_TV_V2.docx";

interface ContractData {
  nomeCompleto: string;
  nomeComercial: string;
  razaoSocial?: string;
  cnpjCpf: string;
  email: string;
  chavePix?: string;
  youtubeId?: string;
  genero?: string;
  estadoCivil?: string;
  cep?: string;
  endereco?: string;
  cidade?: string;
  uf?: string;
  pais?: string;
}

export function generateContractDocxBuffer(data: ContractData): Buffer {
  const templatePath = path.join(process.cwd(), TEMPLATE_RELATIVE_PATH);
  
  if (!fs.existsSync(templatePath)) {
    throw new Error(`Template contract not found at ${templatePath}`);
  }

  const content = fs.readFileSync(templatePath);
  const zip = new PizZip(content);

  // Determine gender/nationality details
  const generoVal = (data.genero || "").toUpperCase();
  let nacionalidade = "brasileiro(a)";
  if (generoVal === "M") {
    nacionalidade = "brasileiro";
  } else if (generoVal === "F") {
    nacionalidade = "brasileira";
  }

  const addressStr = `${data.endereco || ""}, ${data.cidade || ""} - ${data.uf || ""}, CEP ${data.cep || ""}, ${data.pais || "Brasil"}`;

  const isCnpj = data.cnpjCpf.replace(/\D/g, "").length > 11;
  const cpfOrCnpj = isCnpj ? "CNPJ" : "CPF";
  const tipoPessoa = isCnpj ? "jurídica" : "física";
  const nomeParte = data.razaoSocial || data.nomeCompleto;

  const replacements: Record<string, string> = {
    "[nome PESSOA FISICA OU JURÍDICA]": nomeParte.toUpperCase(),
    "[física/jurídica]": tipoPessoa,
    "[CPF/CNPJ]": cpfOrCnpj,
    "[número cpf OU cnpj]": data.cnpjCpf,
    "[NOME]": data.nomeCompleto.toUpperCase(),
    "[ESTADO civil]": (data.estadoCivil || "solteiro(a)").toLowerCase(),
    "[CPF]": data.cnpjCpf,
    "[ENDEREÇO]": addressStr,
  };

  // Files in DOCX zip where text can reside
  const filesToProcess = [
    "word/document.xml",
    "word/header1.xml",
    "word/footer1.xml",
    "word/header2.xml",
    "word/footer2.xml"
  ];

  for (const filename of filesToProcess) {
    const file = zip.file(filename);
    if (file) {
      let text = file.asText();
      
      // Perform replacements
      for (const [key, value] of Object.entries(replacements)) {
        // Try to replace exact match
        text = text.replaceAll(key, value);
      }
      
      zip.file(filename, text);
    }
  }

  return zip.generate({ type: "nodebuffer" });
}
